opds_import.py

```diff
@@ -465,6 +467,74 @@ def data_source(self):
             self._db, self.data_source_name, autocreate=True,
             offers_licenses = offers_licenses
         )
+
+    def assert_importable_content(self, feed, feed_url, max_get_attempts=5):
+        """Raise an exception if the given feed contains nothing that can,
+        even theoretically, be turned into a LicensePool.
+
+        By default, this means the feed must link to open-access content
+        that can actually be retrieved.
+        """
+        metadata, failures = self.extract_feed_data(feed, feed_url)
+        get_attempts = 0
+
+        # Find an open-access link, and try to GET it just to make
+        # sure OPDS feed isn't hiding non-open-access stuff behind an
+        # open-access link.
+        #
+        # To avoid taking forever or antagonizing API providers, we'll
+        # give up after `max_get_attempts` failures.
+        for link in self._open_access_links(metadata.values()):
+            url = link.href
+            success = self._is_open_access_link(url, link.media_type)
+            if success:
+                return success
+            get_attempts += 1
+            if get_attempts >= max_get_attempts:
+                error = "Was unable to GET supposedly open-access content such as %s (tried %s times)" % (
+                    url, get_attempts
+                )
+                explanation = "This might be an OPDS For Distributors feed, or it might require different authentication credentials."
+                raise IntegrationException(error, explanation)
+
+        raise IntegrationException(
+            "No open-access links were found in the OPDS feed.",
+            "This might be an OPDS for Distributors feed."
+        )
+
+    @classmethod
+    def _open_access_links(cls, metadatas):
+        """Find all open-access links in a list of Metadata objects.
+
+        :param metadatas: A list of Metadata objects.
+        :yield: A sequence of `LinkData` objects.
+        """
+        for item in metadatas:
+            if not item.circulation:
+                continue
+            for link in item.circulation.links:
+                if link.rel == Hyperlink.OPEN_ACCESS_DOWNLOAD:
+                    yield link
+
+    def _is_open_access_link(self, url, type):
+        """Is `url` really an open-access link?
+
+        That is, can we make a normal GET request and get something
+        that looks like a book?
+        """
+        headers = {}
+        if type:
+            headers["Accept"] = type
+        status, headers, body = self.http_get(url, headers=headers)
+        if status == 200 and len(body) > 1024 * 10:
```

Comment:
Where does `1024 * 10` come from?

Reply:
I made it up. 10 kilobytes is smaller than any reasonably-sized EPUB but larger than most "uh-oh, there was an error" API error documents.